adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/impl/MessageMapping.java

```diff
@@ -0,0 +1,115 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.adapter.mqtt.impl;
+
+import org.eclipse.hono.adapter.mqtt.MqttContext;
+import org.eclipse.hono.adapter.mqtt.MqttProtocolAdapterProperties;
+import org.eclipse.hono.config.MapperEndpoint;
+import org.eclipse.hono.util.MessageHelper;
+import org.eclipse.hono.util.RegistrationConstants;
+import org.eclipse.hono.util.ResourceIdentifier;
+
+import io.vertx.core.Future;
+import io.vertx.core.MultiMap;
+import io.vertx.core.Promise;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.impl.headers.VertxHttpHeaders;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.client.HttpResponse;
+import io.vertx.ext.web.client.WebClient;
+import io.vertx.mqtt.messages.MqttPublishMessage;
+import io.vertx.mqtt.messages.impl.MqttPublishMessageImpl;
+
+/**
+ * A Vert.x based message mapping component.
+ */
+public final class MessageMapping {
+
+    private WebClient webClient;
+
+    /**
+     * Constructs the messageMapping client used to call external/custom messageMapping.
+     *
+     * @param webClient Vert.x webclient to use in the messageMapping
+     */
+    public MessageMapping(final WebClient webClient) {
+        this.webClient = webClient;
+    }
+
+    /**
+     * Fetches the mapper if configured and calls the external mapping service.
+     *
+     * @param ctx The mqtt context
+     * @param targetAddress The resourceIdentifier with the current targetAddress
+     * @param message Received message
+     * @param registrationInfo information retrieved from the device registry
+     * @param config Mqtt protocol adapter properties
+     * @return Mapped message
+     */
+    public Future<MappedMessage> mapMessage(final MqttContext ctx, final ResourceIdentifier targetAddress,
+                                     final MqttPublishMessage message, final JsonObject registrationInfo, final MqttProtocolAdapterProperties config) {
+        final Promise<MappedMessage> result = Promise.promise();
+        if (!registrationInfo.containsKey(RegistrationConstants.FIELD_MAPPER)) {
+            result.complete(new MappedMessage(ctx, targetAddress, message));
+            return result.future();
+        }
+        final String mapper = registrationInfo.getString(RegistrationConstants.FIELD_MAPPER);
```

Comment:
this will throw a ClassCastException if the property is not a string. IMHO you should use a more defensive approach here ...